File: leetcode/permutations.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<vector<int>> permute(vector<int>& nums)
    {
        vector<vector<int>> res;
        int size = nums.size();

        function<void(int, vector<int>&)> backtrack = [&](int l, vector<int>& cur) {
            if (l >= size) {
                res.push_back(cur);
                return;
            }

            for (auto i = l; i < size; ++i) {
                swap(cur[l], cur[i]);

                backtrack(l + 1, cur);

                swap(cur[l], cur[i]);
            }
        };

        backtrack(0, nums);

        return res;
    }
};

File: leetcode/permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        length = len(nums)

        res = []

        def backtrack(cur: List[int]):
            if len(cur) == length:
                res.append(cur.copy())
                return

            l = nums.copy()
            for a in l:
                nums.pop(0)
                cur.append(a)

                backtrack(cur)

                nums.append(a)
                cur.pop()

        backtrack([])

        return res
